### PR TITLE
Financial Connections: Added support for custom manual entry

### PR DESCRIPTION
**Pull request by kgaidis-stripe**
## Summary

^

Integration/spec doc:
- https://docs.google.com/document/d/1sgRpNTJF4FSnGkWp4unFMS0lgYvqRL2YQpJKjE_NuzA/edit#

## Testing

https://user-images.githubusercontent.com/105514761/216669737-b765eea2-194c-4cb5-bb36-2f62014f996f.mov

Error is received in result:

Screen Shot 2023-02-03 at 12 33 08 PM

Old version still works:

https://user-images.githubusercontent.com/105514761/216670587-e1f048ab-aadd-4b29-b04b-2bfc637e9064.mov

### CODE REVIEW

File: StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift [hunk 1]
```diff
@@ -122,7 +152,9 @@ public extension StripeAPI {
                 accounts: accounts,
                 livemode: try container.decode(Bool.self, forKey: .livemode),
                 paymentAccount: try? container.decode(PaymentAccount.self, forKey: .paymentAccount),
-                bankAccountToken: try? container.decode(BankAccountToken.self, forKey: .bankAccountToken)
+                bankAccountToken: try? container.decode(BankAccountToken.self, forKey: .bankAccountToken),
+                status: (try? container.decode(Status.self, forKey: .status)) ?? .pending,
```
Comment (vardges-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift:156]: @kgaidis-stripe it seems semantically more correct to have the default value as `.unparsable` as it is indeed unparsable in this case.

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/NativeFlowController.swift [hunk 2]
```diff
@@ -226,42 +225,52 @@ extension NativeFlowController {
...
                             )
-                            finishAuthSession(.completed(session: session))
-                        } else if let closeAuthFlowError = closeAuthFlowError {
-                            self.logCompleteEvent(
-                                type: eventType,
-                                status: "failed",
-                                error: closeAuthFlowError
-                            )
-                            finishAuthSession(.failed(error: closeAuthFlowError))
+                            finishAuthSession(.failed(error: FinancialConnectionsSheetError.unknown(debugDescription: "custom manual entry")))
```
Comment (vardges-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/NativeFlowController.swift:241]: So in this case users have to check the text to know it's manual entry?

Comment (kgaidis-stripe) [reply StripeFinancialConnections/StripeFinancialConnections/Source/Native/NativeFlowController.swift:241]: DM'd on this; this was temporary hack (PR in draft)

File: StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift [hunk 3]
```diff
@@ -77,6 +99,8 @@ public extension StripeAPI {
         public let livemode: Bool
         @_spi(STP) public let paymentAccount: PaymentAccount?
         @_spi(STP) public let bankAccountToken: BankAccountToken?
+        let status: Status?
+        let statusDetails: StatusDetails?
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift:103]: backend model to double-check:
- https://livegrep.corp.stripe.com/view/stripe-internal/pay-server/lib/bank_connections/api/resource/link_account_session_resource.rb#L91-106

File: StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift [hunk 4]
```diff
@@ -69,6 +69,28 @@ public extension StripeAPI {
             }
         }
 
+        enum Status: String, SafeEnumCodable, Equatable {
+            case pending
+            case succeeded
+            case failed
+            case cancelled
+            case unparsable
+        }
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift:78]: backend model to double check: 
- https://livegrep.corp.stripe.com/view/stripe-internal/pay-server/lib/bank_connections/api/resource/enums.rb#L11-14

File: StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSessionManifest.swift [hunk 5]
```diff
@@ -45,6 +45,12 @@ struct FinancialConnectionsSessionManifest: Decodable {
         case unparsable
     }
 
+    enum ManualEntryMode: String, SafeEnumCodable, Equatable {
+        case automatic
+        case custom
+        case unparsable
+    }
+
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSessionManifest.swift:53]: backend model to double check:
- https://livegrep.corp.stripe.com/view/stripe-internal/pay-server/lib/bank_connections/api/resource/enums/manual_entry_mode_api_enum.rb#L5-15

File: StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSessionManifest.swift [hunk 6]
```diff
@@ -78,4 +84,5 @@ struct FinancialConnectionsSessionManifest: Decodable {
     let experimentAssignments: [String: String]?
     let assignmentEventId: String?
     let skipSuccessPane: Bool?
+    let manualEntryMode: ManualEntryMode
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSessionManifest.swift:87]: backend model to double-check:
- https://livegrep.corp.stripe.com/view/stripe-internal/pay-server/lib/bank_connections/api/resource/link_account_session_manifest_resource.rb#L22-25

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/NativeFlowController.swift [hunk 7]
```diff
@@ -226,42 +225,54 @@ extension NativeFlowController {
...
+                                        error: terminalError
+                                    )
+                                    finishAuthSession(.failed(error: terminalError))
+                                } else {
+                                    self.logCompleteEvent(
+                                        type: eventType,
+                                        status: "canceled"
+                                    )
+                                    finishAuthSession(.canceled)
+                                }
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/NativeFlowController.swift:275]: all of this should remain unchanged (just if-statement moved the code)

File: StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift [hunk 8]
```diff
@@ -77,6 +99,8 @@ public extension StripeAPI {
         public let livemode: Bool
         @_spi(STP) public let paymentAccount: PaymentAccount?
         @_spi(STP) public let bankAccountToken: BankAccountToken?
+        let status: Status?
```
Comment (vardges-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift:102]: looks like status is non-optional @kgaidis-stripe

Comment (kgaidis-stripe) [reply StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/Models/FinancialConnectionsSession.swift:102]: @vardges-stripe I believe if a merchant is NOT exposed into the flag, it actually returns a NULL (I know, confusing)

**Review by vardges-stripe — APPROVED** [review]

**Comment by kgaidis-stripe** [comment]
@vardges-stripe 

I added a new commit for web flow: https://github.com/stripe/stripe-ios/pull/2261/commits/76f66b165816c6a98f95f6ee7c07f8b034932fde

And there's a video testing it (should be easy to verify the code change by reading too)


https://user-images.githubusercontent.com/105514761/217656598-a767903c-e923-4604-b2ba-f20b565862e8.mov

**Pull request merged**